File: src/main/java/benchmark/JMHTest.java
package benchmark;

import org.openjdk.jmh.annotations.*;

import java.util.concurrent.TimeUnit;

/**
 * 比较字符串直接相加和StringBuilder的效率
 *
 * @author XiongNeng
 * @version 1.0
 * @since 2018/1/7
 */
@BenchmarkMode(Mode.AverageTime)
@Warmup(iterations = 3)
@Measurement(iterations = 10, time = 5, timeUnit = TimeUnit.SECONDS)
@Threads(8)
@Fork(2)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
public class JMHTest {

    @Benchmark
    public void testStringAdd() {
//        String a = "";
        for (int i = 0; i < 10000000; i++) {
            int a  = 1 + i;
        }
//        print(a);
    }

    @Benchmark
    public void testStringBuilderAdd() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 10; i++) {
            sb.append(i);
        }
        print(sb.toString());
    }

    private void print(String a) {
    }
}

File: src/main/scala/zpj/jmh/JMHTest.scala
package zpj.jmh

import java.util.concurrent.TimeUnit

import org.openjdk.jmh.annotations._

import scalaz.std.java.time

/**
  * Created by PerkinsZhu on 2018/6/23 9:58
  **/
@BenchmarkMode(Array(Mode.Throughput))
@Warmup(iterations = 3)
@Measurement(iterations = 10, time = 5, timeUnit = TimeUnit.SECONDS)
@Threads(8)
@Fork(2)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
class JMHTestScala {
  @Benchmark
  def test(): Unit ={
    1 to 10000 foreach(i => i * i)
  }

}
